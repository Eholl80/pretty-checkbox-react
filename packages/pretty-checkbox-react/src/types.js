// @flow

import * as React from 'react';

export type PCRColors = 'primary' | 'success' | 'info' | 'warning' | 'danger';
export type PCRColorsOutline = 'primary-o' | 'success-o' | 'info-o' | 'warning-o' | 'danger-o';

export type RadioCheckboxVariants = 'curve' | 'round' | 'plain';
export type RadioCheckboxFillModes = 'fill' | 'thick';
export type SwitchFillModes = 'outline' | 'fill' | 'slim';

// IconType not applicable to Switch
export type IconPropType = {
    type: 'svg' | 'image' | 'icon' | null,
    icon: ?React.Element<*>,
};

export type SwitchAnimations = 'smooth' | 'jelly' | 'tada';
export type RadioCheckboxAnimations = SwitchAnimations | 'rotate' | 'pulse';
