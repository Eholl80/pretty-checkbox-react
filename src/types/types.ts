export type PCRColors = 'primary' | 'success' | 'info' | 'warning' | 'danger';
export type PCRColorsOutline = 'primary-o' | 'success-o' | 'info-o' | 'warning-o' | 'danger-o';

export type RadioCheckboxVariants = 'curve' | 'round' | 'plain';
export type RadioCheckboxFillModes = 'fill' | 'thick';
export type SwitchFillModes = 'outline' | 'fill' | 'slim';

export type SwitchAnimations = 'smooth' | 'jelly' | 'tada';
export type RadioCheckboxAnimations = SwitchAnimations | 'rotate' | 'pulse';
