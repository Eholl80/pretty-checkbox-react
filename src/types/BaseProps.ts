import * as React from 'react';

import {
    PCRColors,
    PCRColorsOutline,
    RadioCheckboxVariants,
    RadioCheckboxFillModes,
    RadioCheckboxAnimations,
    SwitchFillModes,
} from './types';

export type BaseProps<S> = React.HTMLAttributes<HTMLInputElement> & {
    /**
     * The current checkbox state.
     */
    state: S;

    /**
     * Dispatch method from React.useState or `this.setState` from class component.
     */
    setState: React.Dispatch<React.SetStateAction<S>>;

    /**
     * A label to show in the Checkbox. This can be any JSX node or string.
     */
    children?: React.ReactNode;

    /**
     * Specify one of the colors.
     */
    color?: PCRColors | PCRColorsOutline;

    /**
     * Specify one of the variants.
     */
    variant?: RadioCheckboxVariants;

    /**
     * Specify one of the fill modes.
     */
    fill?: RadioCheckboxFillModes | SwitchFillModes;

    /**
     * Specify one of the animations.
     */
    animation?: RadioCheckboxAnimations;

    /**
     * Pass an icon to render.
     */
    icon?: React.ReactComponentElement<any, HTMLElement>;

    /**
     * Runs inside the default onChange handler for the Checkbox.
     */
    onChange?: (e: React.ChangeEvent<HTMLInputElement>) => void;

    /**
     * Set the checkbox as disabled.
     */
    disabled?: boolean;

    /**
     * Set the checkbox as locked. This will also set the
     * `aria-disabled` attribute to `true`.
     */
    locked?: boolean;

    /**
     * Pass additional class names to the `.pretty` div.
     */
    className?: string;

    /**
     * Set the value of the input field.
     */
    value?: string;

    /**
     * Make the checkbox, radio, or switcher bigger.
     */
    bigger?: boolean;
};
