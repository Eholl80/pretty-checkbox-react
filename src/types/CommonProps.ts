import { InputHTMLAttributes } from 'react';

type Colors = 'primary' | 'success' | 'warning' | 'info' | 'danger';
type ColorsOutline = 'primary-o' | 'success-o' | 'warning-o' | 'info-o' | 'danger-o';

type CheckboxRadioProps = {
    icon?: React.ReactElement<any, 'img' | 'svg' | 'i'>;
    iconType?: 'icon' | 'svg' | 'image';
    fill?: 'fill' | 'thick';
    shape?: 'square' | 'curve' | 'round';
    animation?: 'smooth' | 'jelly' | 'tada' | 'rotate' | 'pulse';
    plain?: boolean;
};

export type SwitchShape = 'fill' | 'outline' | 'slim';

export type CommonProps<S> = InputHTMLAttributes<HTMLInputElement> & {
    state: S;
    setState?: React.Dispatch<React.SetStateAction<S>>;
    onChange: (e: React.ChangeEvent<HTMLInputElement>, value?: any) => void;
    color?: Colors | ColorsOutline;
    locked?: boolean;
    bigger?: boolean;
    className?: string;
    type?: 'checkbox' | 'radio';
    baseId?: string;
};

export type CommonCheckboxRadioProps<S = any> = CommonProps<S> & CheckboxRadioProps;
