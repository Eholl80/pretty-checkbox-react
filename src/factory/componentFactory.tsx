import * as React from 'react';
import classNames from 'classnames';

import { getChecked } from './utils';
import { CommonCheckboxRadioProps } from '../types/CommonProps';

export const createInput = ({
    onChange,
    disabled,
    value,
    state,
    locked,
    type,
    name,
    checked,
}: CommonCheckboxRadioProps) => {
    return (
        <input
            onChange={onChange}
            type={type}
            value={value}
            name={name}
            tabIndex={-1}
            checked={
                typeof checked === 'undefined'
                    ? getChecked({ state, value, disabled, locked, type })
                    : checked
            }
            disabled={disabled}
        />
    );
};

export const createLabel = ({ color, icon, children }: CommonCheckboxRadioProps) => (
    <div
        className={classNames('state', {
            [`p-${color}`]: color,
        })}>
        {icon}
        <label>{children}</label>
    </div>
);
