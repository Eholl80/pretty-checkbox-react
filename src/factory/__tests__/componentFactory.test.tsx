import { createInput, createLabel } from '../componentFactory';
import { CommonCheckboxRadioProps } from '../../types/CommonProps';

describe('componentFactory tests', () => {
    describe('createInput tests', () => {
        const props: CommonCheckboxRadioProps = {
            onChange: jest.fn(),
            disabled: false,
            locked: false,
            state: false,
            value: undefined,
            type: 'checkbox',
            name: undefined,
            checked: false,
        };

        it('should render without errors', () => {
            expect(() => {
                createInput(props);
            }).not.toThrow();
        });

        it('should match the snapshot', () => {
            expect(createInput(props)).toMatchSnapshot();
        });

        it('should be checked when rendered', () => {
            expect(
                createInput({
                    ...props,
                    state: 'abc',
                    checked: undefined,
                })
            ).toMatchSnapshot();
        });
    });

    describe('createLabel tests', () => {
        const props: CommonCheckboxRadioProps = {
            state: false,
            onChange: jest.fn(),
            icon: undefined,
            children: 'testing',
        };

        it('should render without errors', () => {
            expect(() => {
                createLabel(props);
            }).not.toThrow();
        });

        it('should match the snapshot', () => {
            expect(createLabel(props)).toMatchSnapshot();
        });
    });
});
