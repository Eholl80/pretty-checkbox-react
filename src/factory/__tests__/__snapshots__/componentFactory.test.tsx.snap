// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`componentFactory tests createInput tests should be checked when rendered 1`] = `
<input
  checked={true}
  disabled={false}
  onChange={[MockFunction]}
  tabIndex={-1}
  type="checkbox"
/>
`;

exports[`componentFactory tests createInput tests should match the snapshot 1`] = `
<input
  checked={false}
  disabled={false}
  onChange={[MockFunction]}
  tabIndex={-1}
  type="checkbox"
/>
`;

exports[`componentFactory tests createLabel tests should match the snapshot 1`] = `
<div
  className="state"
>
  <label>
    testing
  </label>
</div>
`;
